src/abstractdataframe/iteration.jl

```diff
@@ -27,25 +27,32 @@ with each row represented as a `DataFrameRow`.
 """
 eachrow(df::AbstractDataFrame) = DFRowIterator(df)
 
+Base.size(itr::DFRowIterator) = (size(itr.df, 1), )
+Base.size(itr::DFRowIterator, ix) =
+    ix == 1 ? length(itr) : throw(ArgumentError("Incorrect dimension"))
+Base.length(itr::DFRowIterator) = size(itr.df, 1)
+Base.firstindex(itr::DFRowIterator) = 1
+Base.lastindex(itr::DFRowIterator) = length(itr)
+
 function Base.iterate(itr::DFRowIterator, i=1)
     i > size(itr.df, 1) && return nothing
     return (DataFrameRow(itr.df, i), i + 1)
 end
+
 Base.eltype(::DFRowIterator{T}) where {T} = DataFrameRow{T}
-Base.size(itr::DFRowIterator) = (size(itr.df, 1), )
-Base.length(itr::DFRowIterator) = size(itr.df, 1)
 Base.getindex(itr::DFRowIterator, i) = DataFrameRow(itr.df, i)
 
 # Iteration by columns
 """
-    DFColumnIterator{<:AbstractDataFrame}
+    DFColumnIterator{<:AbstractDataFrame, C}
 
 Iterator over columns of an `AbstractDataFrame`.
-Each returned value is a tuple consisting of column name and column vector.
-
-A value of this type is returned by the [`eachcol`](@link) function.
+If `C` is `true` (a value returned by the [`eachcol`](@link) function)
+then each returned value is a tuple consisting of column name and column vector.
+If `C` is `false` (a value returned by the [`columns`](@link) function)
+then each returned value is a column vector.
 """
-struct DFColumnIterator{T <: AbstractDataFrame}
+struct DFColumnIterator{T<:AbstractDataFrame, C}
```

Comment:
Instead of `C` being a Boolean, it could be either `AbstractVector` or `Tuple{Symbol,AbstractVector}`. That way you could do `<: AbstractArray{C}` here, and you wouldn't need to define `eltype` and such below.

BTW, it's a good occasion to move from a tuple to a pair. That shouldn't affect most uses but it's more natural here.

Reply:
That said, if we keep the special behavior of `map` which returns a `DataFrame`, maybe we'd better not make this iterator an `AbstractVector`. Same for the row iterator if we want to do something similar in the future.

Reply:
OK, so we need to choose between these two solutions:
1. Define `map(f, columns(df))` to return a data frame. We could add `map(f, rows(df))` later if we want and that would be consistent. People have to use comprehensions to get a vector instead.
2. Keep the standard behavior of `map(f, columns(df))`, i.e. have it return a vector. This is basically `colwise`, but with a more standard name (and we could deprecate it, the broadcasting form `f.(columns(df))` is even more compact). We need another way to apply a function to each column of a data frame and get a data frame back: maybe `mapcols(f, df)`, and `mapcols!(f, df)` for the in-place variant (the naming would be consistent with other column-wise functions, cf. https://github.com/JuliaData/DataFrames.jl/issues/1514). Then `map(f, df)` would apply a function to each row (this is what JuliaDB [supports](http://juliadb.org/latest/api/selection.html#Map-1)).

> Also, I do not see much sense of doing the same functionality on `eachrow` (we would not know how to transform it to a `DataFrame` later).

That would be quite simple AFAICT: require the user-provided function to return a named tuple, and fill a data frame with that like we now do in `map(::Function, ::GroupedDataFrame)`.

Cc: @piever to discuss the consistency with JuliaDB?

Reply:
For me to understand, there are a `ColumnIterator` and `RowIterator` types, and the decision is whether one should overload `map` on them to return a `DataFrame` rather than a vector of columns or a vector of rows?

In JuliaDB `map` acts by default on the rows and returns a `table`. We also have a `map_rows` which basically would correspond to `map_rows(f, iter...) = DataFrame(f(i) for i in iter...)`. It is useful when `f` returns `NamedTuples` but `iter...` are not a `DataFrame` (they could be vectors or iterators), for example `map_rows((x, y) -> (s = x+y, d = x-y), rand(100), rand(100))`. Could this be added to `DataFrames` as well? In which case I think `mapcols` `map_rows` is a strange pair. What could be a better pair of names?

For consistency with JuliaDB, option 2 seems much better (also, to keep `map` and `filter` consistent). I thought there was a `colwise` or similar in JuliaDB but apparently I'm misremembering. It looks like currently one has to do:

```julia
julia> t = table((a = rand(100), b = randn(100)));

julia> f(v) = partialsort(v, 1:10)
f (generic function with 1 method)

julia> table(map(f, columns(t)))
Table with 10 rows, 2 columns:
a           b
────────────────────
0.00392291  -2.52979
0.0221383   -2.01489
0.029867    -1.92054
0.0462468   -1.89799
0.0720458   -1.5951
0.0824684   -1.31414
0.105048    -1.25503
0.108803    -1.19262
0.1204      -1.13395
0.136053    -1.13125
```

This I think is not so bad but maybe it'd be more awkward in DataFrames because `DataFrame(map(f, columns(t)))` would lose the column names?

Reply:
> In JuliaDB `map` acts by default on the rows and returns a `table`. We also have a `map_rows` which basically would correspond to `map_rows(f, iter...) = DataFrame(f(i) for i in iter...)`. It is useful when `f` returns `NamedTuples` but `iter...` are not a `DataFrame` (they could be vectors or iterators), for example `map_rows((x, y) -> (s = x+y, d = x-y), rand(100), rand(100))`. Could this be added to `DataFrames` as well? In which case I think `mapcols` `map_rows` is a strange pair. What could be a better pair of names?

I don't really see the advantage of providing a `map_rows` function if `DataFrame(f(i) for i in iter...)` does the same thing TBH. And using a generator like that sounds more Julian (it's like creating dicts).

> For consistency with JuliaDB, option 2 seems much better (also, to keep `map` and `filter` consistent).

OK. That's good since that's the approach @bkamins has implemented. ;-)

But the example you show below goes a bit against with this statement AFAICT: `map(f, columns(t))` returns a `NamedTuple` in JuliaDB, not a vector (which is why names are lost). That's logical since `column(t)` gives a `NamedTuple`, but that would be different from data frames. That said, it's not a big deal. Maybe JuliaDB could support `mapcols` just for consistency and convenience?

Reply:
Yep, maybe JuliaDB could deprecate `map_rows` as "unnecessary". The easiest is probably to add a `mapcols(f, t; select = All())` method to JuliaDB. As I mentioned above, I'm afraid this creates a bit of cacophony on the JuliaDB side (`mapcols` and `map_rows` have different underscore style and are not related in the obvious way). The cleanest is probably to deprecate `map_rows` and rename `setcol`, `popcol`, `pushcol`... to `setcols`, `popcols`, `pushcols` (as they accept multiple columns at once) and add `mapcols` to the collections. But this PR shouldn't worry about that as all the changes to make things consistent should come from JuliaDB.

Reply:
Makes sense. It's good to anticipate to what API we both want to converge in the long term.